### PR TITLE
Add ownership check to E2EFilterTest

### PR DESCRIPTION
**Pull request by oerling**
Makes a shallow and a deep copy of every even numbered batch in the
different tests. Checks that these remain =equal after receiving the
odd-numbered batch. Does this for the first 12 batches of each test
run so as not to increase test time.

This effectively checks that when the reader hands out a second
reference to a Buffer, this does not get overwritten if the Buffer is
multiply referenced when getting the next batch. The same applies to
reusing parts of the 'result' VectorPtr in getValues of the different
reader classes.

Fixes a bad reuse of a multiply referenced RowVector child of the
top-level RowVector. Note that RowVector children of RowVectors must
be initialized to at least an empty RowVector, while other children
can be set to nullptr.

### CODE REVIEW

File: velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp [hunk 1]
```diff
@@ -193,13 +193,43 @@ void SelectiveStructColumnReader::read(
...
       *result != nullptr,
       "SelectiveStructColumnReader expects a non-null result");
-  RowVector* resultRow = dynamic_cast<RowVector*>(result->get());
+  auto resultRow = dynamic_cast<RowVector*>(result->get());
   VELOX_CHECK(resultRow, "Struct reader expects a result of type ROW.");
+
+  if (!result->unique()) {
+    std::vector<VectorPtr> children(resultRow->children().size());
+    fillRowVectorChildren(
+        *resultRow->pool(), resultRow->type()->as<TypeKind::ROW>(), children);
```
Comment (mbasmanova) [inline velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp:224]: nit: resultRow->type()->asRow()

File: velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp [hunk 2]
```diff
@@ -193,13 +193,43 @@ void SelectiveStructColumnReader::read(
...
+    *result = std::make_unique<RowVector>(
+        resultRow->pool(),
+        resultRow->type(),
+        BufferPtr(nullptr),
+        0,
+        std::move(children));
+    resultRow = reinterpret_cast<RowVector*>(result->get());
```
Comment (mbasmanova) [inline velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp:231]: why reinterpret_cast and not static_cast?

File: velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp [hunk 3]
```diff
@@ -193,13 +193,43 @@ void SelectiveStructColumnReader::read(
...
+}
+} // namespace
+
 void SelectiveStructColumnReader::getValues(RowSet rows, VectorPtr* result) {
   assert(!children_.empty());
   VELOX_CHECK(
       *result != nullptr,
       "SelectiveStructColumnReader expects a non-null result");
-  RowVector* resultRow = dynamic_cast<RowVector*>(result->get());
+  auto resultRow = dynamic_cast<RowVector*>(result->get());
```
Comment (mbasmanova) [inline velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp:218]: nit: this can be a static cast if the check is moved ahead of this and changed to result->type()->isRow()

File: velox/dwio/dwrf/test/E2EFilterTestBase.h [hunk 4]
```diff
@@ -273,6 +286,15 @@ class E2EFilterTestBase : public testing::Test {
   int32_t flushEveryNBatches_{10};
   int32_t nextReadSizeIndex_{0};
   std::vector<int32_t> readSizes_;
+
+  // Sequence number of batch for checkOwnership().
+  int32_t batchCounter_;
```
Comment (mbasmanova) [inline velox/dwio/dwrf/test/E2EFilterTestBase.h:291]: Would it make sense to move these into a helper class OwnershipChecker for modularity?

File: velox/dwio/dwrf/test/E2EFilterTestBase.cpp [hunk 5]
```diff
@@ -411,4 +394,33 @@ void E2EFilterTestBase::testWithTypes(
...
+  }
+  if (batchCounter_ % 2 == 0) {
+    previousBatch_ = std::make_shared<RowVector>(
+        batch->pool(),
+        batch->type(),
+        BufferPtr(nullptr),
+        batch->size(),
+        batch->as<RowVector>()->children());
+    previousBatchCopy_ =
+        BaseVector::create(batch->type(), batch->size(), batch->pool());
```
Comment (mbasmanova) [inline velox/dwio/dwrf/test/E2EFilterTestBase.cpp:412]: It would be nice to add a helper static method BaseVector::copy(const VectorPtr& vector)

File: velox/dwio/dwrf/test/E2EFilterTestBase.cpp [hunk 6]
```diff
@@ -144,39 +144,26 @@ void E2EFilterTestBase::makeStringUnique(const Subfield& field) {
 
 void E2EFilterTestBase::makeNotNull(int32_t firstRow) {
   for (RowVectorPtr batch : batches_) {
```
Comment (mbasmanova) [inline velox/dwio/dwrf/test/E2EFilterTestBase.cpp:146]: RowVectorPtr -> const auto&

**Review by mbasmanova — COMMENTED** [review]
@oerling This type of testing is super valuable. I wonder if this test would fail without a fix from #1483 - Fix reuse of nullsInReadRange_ in the DWRF reader. Would you check to confirm that the test is able to catch a range of memory re-use problems?

**Comment by oerling** [comment]
Updated the PR.

File: velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp [hunk 7]
```diff
@@ -193,13 +193,51 @@ void SelectiveStructColumnReader::read(
...
-  VELOX_CHECK(resultRow, "Struct reader expects a result of type ROW.");
+  VELOX_CHECK(
+      result->get()->type()->isRow(),
+      "Struct reader expects a result of type ROW.");
+
+  auto resultRow = static_cast<RowVector*>(result->get());
+  if (!result->unique()) {
+    std::vector<VectorPtr> children(resultRow->children().size());
+    fillRowVectorChildren(
+        *resultRow->pool(), resultRow->type()->as<TypeKind::ROW>(), children);
```
Comment (mbasmanova) [inline velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp:232]: nit: you can use asRow() here as well

File: velox/dwio/dwrf/test/E2EFilterTestBase.h [hunk 8]
```diff
@@ -57,6 +57,27 @@ inline void TestingHook<StringView>::addValue(
...
+// Utility for checking that a subsequent batch of output does not
+// overwrite internals of a possibly retained previous batch.
+class OwnershipChecker {
+ public:
+  // Receives consecutive batches during a test and checks that
+  // subsequent ones do not overwrite retained content from previous
+  // ones.
+
+  void check(VectorPtr batch);
+  // Sequence number of batch for checkOwnership().
```
Comment (mbasmanova) [inline velox/dwio/dwrf/test/E2EFilterTestBase.h:69]: This comment is not followed by any code. Please, check.

File: velox/dwio/dwrf/test/E2EFilterTestBase.h [hunk 9]
```diff
@@ -57,6 +57,27 @@ inline void TestingHook<StringView>::addValue(
...
 }
 
+// Utility for checking that a subsequent batch of output does not
+// overwrite internals of a possibly retained previous batch.
+class OwnershipChecker {
+ public:
+  // Receives consecutive batches during a test and checks that
+  // subsequent ones do not overwrite retained content from previous
+  // ones.
+
```
Comment (mbasmanova) [inline velox/dwio/dwrf/test/E2EFilterTestBase.h:67]: empty line between comment and method; please, remove

File: velox/dwio/dwrf/test/E2EFilterTestBase.h [hunk 10]
```diff
@@ -262,6 +283,8 @@ class E2EFilterTestBase : public testing::Test {
     return readSizes_[nextReadSizeIndex_++];
   }
 
+  // Called at the start of a series of reads from a RowReader to
```
Comment (mbasmanova) [inline velox/dwio/dwrf/test/E2EFilterTestBase.h:286]: This comment should be removed, I think. Please, check.

File: velox/vector/BaseVector.h [hunk 11]
```diff
@@ -391,6 +391,14 @@ class BaseVector {
     });
   }
 
+  // Utility for making a deep copy of a whole vector.
+  static std::shared_ptr<BaseVector> copy(const BaseVector& vector) {
```
Comment (mbasmanova) [inline velox/vector/BaseVector.h:395]: Nice. Thank you!

File: velox/vector/tests/VectorTest.cpp [hunk 12]
```diff
@@ -419,6 +420,10 @@ class VectorTest : public testing::Test {
       }
       sourcePtr = sourcePtr->valueVector().get();
     }
+    if (!isLazyNotLoaded(*source)) {
```
Comment (mbasmanova) [inline velox/vector/tests/VectorTest.cpp:423]: This change seems unrelated. What is this for?

**Review by mbasmanova — APPROVED** [review]
@oerling Thank you for addressing comments. Looks great % some mixups with the comments.

**Comment by mbasmanova** [comment]
@oerling CirleCI is red. Please, take a look.
